### ISSUE TITLE
fatal error: concurrent map iteration and map write

### ISSUE THREAD

**Issue by ngbrown** [minio/minio#10426]
With cache enabled, on Windows, with the newest version, `minio.RELEASE.2020-09-05T07-14-49Z`, minio crashes with `fatal error: concurrent map iteration and map write`.

## Expected Behavior
Run without crashing.

## Current Behavior
Under some load of uploading multiple files, using multipart uploads, it will crash after some time (less than 30 minutes).

```
fatal error: concurrent map iteration and map write

goroutine 23463 [running]:
runtime.throw(0x1d8fbcd, 0x26)
	runtime/panic.go:1116 +0x79 fp=0xc000523af0 sp=0xc000523ac0 pc=0x4386f9
runtime.mapiternext(0xc000523bd0)
	runtime/map.go:853 +0x559 fp=0xc000523b70 sp=0xc000523af0 pc=0x410659
github.com/minio/minio/cmd.getMetadata(0xc000f3eac0, 0x7, 0xc000f3eac8, 0x31, 0x186fc0a8, 0xed6e6dd42, 0x3575980, 0x24482, 0x0, 0xc0027acc30, ...)
	github.com/minio/minio/cmd/disk-cache.go:178 +0x1e0 fp=0xc000523c40 sp=0xc000523b70 pc=0x1531fb0
github.com/minio/minio/cmd.(*cacheObjects).GetObjectNInfo.func2(0xc000331d60, 0x21619a0, 0xc000e946c0, 0xc000f3eac0, 0x7, 0xc000f3eac8, 0x31, 0xc0005e8030, 0xc000e0e000, 0xc0005e8038)
	github.com/minio/minio/cmd/disk-cache.go:341 +0xfc fp=0xc000523f90 sp=0xc000523c40 pc=0x16fe91c
runtime.goexit()
	runtime/asm_amd64.s:1373 +0x1 fp=0xc000523f98 sp=0xc000523f90 pc=0x469c81
created by github.com/minio/minio/cmd.(*cacheObjects).GetObjectNInfo
	github.com/minio/minio/cmd/disk-cache.go:337 +0xa69
```

## Possible Solution
N/A

## Steps to Reproduce (for bugs)

1. Enable caching on Windows.
2. Upload many files via multi-part upload.
3. Mix-in some downloads.

The `mc admin obd` command fails to download a report, but doesn't crash minio.
Running the warp benchmark locally and on the server didn't cause the fault.

## Context
N/A

## Regression
Yes.

It was working for months with a local build of 2af3004409f46e083855451c9b97f592d8fb1263. I also tested on the next released version, `RELEASE.2020-06-01T17-28-03Z`, which did work. `RELEASE.2020-07-02T00-15-09Z` also worked.

A more recent build, `RELEASE.2020-08-08T04-50-06Z` had the same fatal error with a similar call-stack.

## Your Environment

* Version used (`minio --version`): RELEASE.2020-09-05T07-14-49Z
* Server setup and configuration: The object storage is on a network drive and the remote file-system does occasionally error/timeout with access denied or file not found when completing the multi-part uploads. Retrying always worked and this was not crashing minio.
* Operating System and version (`uname -a`): Windows Server 2016

Hopefully the error log is enough to track the issue down: [minio-service.err.log](https://github.com/minio/minio/files/5179820/minio-service.err.log)

Nothing unexpected occurred in the normal output log: [minio-service.out.log](https://github.com/minio/minio/files/5179835/minio-service.out.log)

**Comment by harshavardhana** [comment]
Thanks @ngbrown looks a regression.. we will take a look

**Comment by ngbrown** [comment]
In the fault log for both `RELEASE.2020-08-08T04-50-06Z` and `RELEASE.2020-09-05T07-14-49Z` another goroutine was in the middle of `(*FSObjects).putObject` (`fs-v1.go`). In both cases, it has just performed this update on the `objInfo.UserDefined` map:

https://github.com/minio/minio/blob/a9aaea0d674389ef72686ab7783ded02ddf010f6/cmd/fs-v1.go#L1212

Looking at the diff between https://github.com/minio/minio/compare/648cb13e023e49af97f07f164a50f0e037ca20b1...0b8255529a6f74df81f975869b6f2b93ff4b4dda, I don't see any obvious changes that would start triggering a concurrency exception.

The fault log for `RELEASE.2020-08-08T04-50-06Z` is here: [minio-service.err2.log](https://github.com/minio/minio/files/5180028/minio-service.err2.log)

**Comment by harshavardhana** [comment]
This bug has existed for long, just that it might have been masked before or rare.

**Comment by ngbrown** [comment]
This is still an issue on `RELEASE.2020-09-08T23-05-18Z`.

The error log contains two fatal errors, since I restarted the service right after it faulted the first time, and it crashed again within minutes.

[minio-service.err3.log](https://github.com/minio/minio/files/5200958/minio-service.err3.log)

**Comment by klauspost** [comment]
Let me take a look.

**Comment by siriusckx** [comment]
this error persists on `minio version RELEASE.2020-12-26T01-35-54Z`

```
API: SYSTEM()
Time: 10:06:20 CST 12/29/2020
DeploymentID: 3ab46ab8-ab4c-40d1-9f2e-3e96f891cdb9
Error: cache version deprecated (will autoupdate) (*errors.errorString)
       2: cmd/data-usage-cache.go:485:cmd.(*dataUsageCache).load()
       1: cmd/erasure.go:361:cmd.erasureObjects.crawlAndGetDataUsage.func2()
fatal error: concurrent map iteration and map write 

goroutine 8930383 [running]:
runtime.throw(0x1f508c0, 0x26) 
    runtime/panic.go:1116 +0x72 fp=0xc0027df628 sp=0xc0027df5f8 pc=0x437952
runtime.mapiternext(0xc0027df790)
    runtime/map.go:853 +0x554 fp=0xc0027df6a8 sp=0xc0027df628 pc=0x410db4
github.com/minio/minio/cmd.(*dataUsageCache).flatten(0xc0027dfe88, 0x3753e42, 0x0, 0x0, 0x0, 0x0, 0x235, 0x32, 0x1fb, 0x8, ...)
    github.com/minio/minio/cmd/data-usage-cache.go:323 +0xe9 fp=0xc0027df870 sp=0xc0027df6a8 pc=0x1658729
github.com/minio/minio/cmd.(*dataUsageCache).dui(0xc0027dfe88, 0x1f03e16, 0x1, 0xc01ba02000, 0x8027, 0x8999, 0x0, 0x0, 0x0, 0x0, ...)
    github.com/minio/minio/cmd/data-usage-cache.go:230 +0xe5 fp=0xc0027dfab8 sp=0xc0027df870 pc=0x16575c5
github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3.1()
    github.com/minio/minio/cmd/erasure-server-sets.go:396 +0x325 fp=0xc0027dfcf8 sp=0xc0027dfab8 pc=0x187fbc5
github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3(0xc002e7f3a8, 0xc00721cd20, 0xc00720e480, 0xc01ba02000, 0x8027, 0x8999, 0xc0072211f0, 0xc01a20cde0, 0xc007221210, 0xc001923140)
    github.com/minio/minio/cmd/erasure-server-sets.go:415 +0x183 fp=0xc0027dff90 sp=0xc0027dfcf8 pc=0x187fee3
runtime.goexit()
    runtime/asm_amd64.s:1374 +0x1 fp=0xc0027dff98 sp=0xc0027dff90 pc=0x470701
created by github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage
    github.com/minio/minio/cmd/erasure-server-sets.go:374 +0x566
```

**Comment by harshavardhana** [comment]
> this error persists on `minio version RELEASE.2020-12-26T01-35-54Z`
> 
> ```
> API: SYSTEM()
> Time: 10:06:20 CST 12/29/2020
> DeploymentID: 3ab46ab8-ab4c-40d1-9f2e-3e96f891cdb9
> Error: cache version deprecated (will autoupdate) (*errors.errorString)
>        2: cmd/data-usage-cache.go:485:cmd.(*dataUsageCache).load()
>        1: cmd/erasure.go:361:cmd.erasureObjects.crawlAndGetDataUsage.func2()
> fatal error: concurrent map iteration and map write 
> 
> goroutine 8930383 [running]:
> runtime.throw(0x1f508c0, 0x26) 
>     runtime/panic.go:1116 +0x72 fp=0xc0027df628 sp=0xc0027df5f8 pc=0x437952
> runtime.mapiternext(0xc0027df790)
>     runtime/map.go:853 +0x554 fp=0xc0027df6a8 sp=0xc0027df628 pc=0x410db4
> github.com/minio/minio/cmd.(*dataUsageCache).flatten(0xc0027dfe88, 0x3753e42, 0x0, 0x0, 0x0, 0x0, 0x235, 0x32, 0x1fb, 0x8, ...)
>     github.com/minio/minio/cmd/data-usage-cache.go:323 +0xe9 fp=0xc0027df870 sp=0xc0027df6a8 pc=0x1658729
> github.com/minio/minio/cmd.(*dataUsageCache).dui(0xc0027dfe88, 0x1f03e16, 0x1, 0xc01ba02000, 0x8027, 0x8999, 0x0, 0x0, 0x0, 0x0, ...)
>     github.com/minio/minio/cmd/data-usage-cache.go:230 +0xe5 fp=0xc0027dfab8 sp=0xc0027df870 pc=0x16575c5
> github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3.1()
>     github.com/minio/minio/cmd/erasure-server-sets.go:396 +0x325 fp=0xc0027dfcf8 sp=0xc0027dfab8 pc=0x187fbc5
> github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3(0xc002e7f3a8, 0xc00721cd20, 0xc00720e480, 0xc01ba02000, 0x8027, 0x8999, 0xc0072211f0, 0xc01a20cde0, 0xc007221210, 0xc001923140)
>     github.com/minio/minio/cmd/erasure-server-sets.go:415 +0x183 fp=0xc0027dff90 sp=0xc0027dfcf8 pc=0x187fee3
> runtime.goexit()
>     runtime/asm_amd64.s:1374 +0x1 fp=0xc0027dff98 sp=0xc0027dff90 pc=0x470701
> created by github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage
>     github.com/minio/minio/cmd/erasure-server-sets.go:374 +0x566
> ```

This should be sufficiently fixed in #11620

**Comment by ToLToL** [comment]
Hi @harshavardhana,

I have exactly the same error on `2021-02-19T04:38:02Z`:

```
fatal error: concurrent map iteration and map write

goroutine 2389494166 [running]:
runtime.throw(0x20214ea, 0x26)
	runtime/panic.go:1116 +0x72 fp=0xc01d10f628 sp=0xc01d10f5f8 pc=0x437952
runtime.mapiternext(0xc01d10f790)
	runtime/map.go:853 +0x554 fp=0xc01d10f6a8 sp=0xc01d10f628 pc=0x410db4
github.com/minio/minio/cmd.(*dataUsageCache).flatten(0xc01d10fe88, 0xa7a5c084cec, 0x0, 0x0, 0x0, 0x0, 0x8b86, 0x9e3, 0x2430, 0x48ce, ...)
	github.com/minio/minio/cmd/data-usage-cache.go:342 +0xe9 fp=0xc01d10f870 sp=0xc01d10f6a8 pc=0x16770c9
github.com/minio/minio/cmd.(*dataUsageCache).dui(0xc01d10fe88, 0x1fd3916, 0x1, 0xc021b98000, 0x2cc0, 0x3733, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/minio/minio/cmd/data-usage-cache.go:249 +0xe5 fp=0xc01d10fab8 sp=0xc01d10f870 pc=0x1675f65
github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3.1()
	github.com/minio/minio/cmd/erasure-server-pool.go:443 +0x325 fp=0xc01d10fcf8 sp=0xc01d10fab8 pc=0x18b06a5
github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3(0xc011b9ee98, 0xc023537e40, 0xc01f548fc0, 0xc021b98000, 0x2cc0, 0x3733, 0xc0017932f0, 0xc0125013e0, 0xc001793310, 0xc001698180)
	github.com/minio/minio/cmd/erasure-server-pool.go:462 +0x183 fp=0xc01d10ff90 sp=0xc01d10fcf8 pc=0x18b09c3
runtime.goexit()
	runtime/asm_amd64.s:1374 +0x1 fp=0xc01d10ff98 sp=0xc01d10ff90 pc=0x470701
created by github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage
	github.com/minio/minio/cmd/erasure-server-pool.go:421 +0x56f
```

It should be fixed in  #11620 on `RELEASE.2021-02-24T18-44-45Z` right?

**Comment by ToLToL** [comment]
Cache is disabled btw, didn't set `MINIO_CACHE="on"`.

**Comment by harshavardhana** [comment]
> Hi @harshavardhana,
> 
> I have exactly the same error on `2021-02-19T04:38:02Z`:
> 
> ```
> fatal error: concurrent map iteration and map write
> 
> goroutine 2389494166 [running]:
> runtime.throw(0x20214ea, 0x26)
> 	runtime/panic.go:1116 +0x72 fp=0xc01d10f628 sp=0xc01d10f5f8 pc=0x437952
> runtime.mapiternext(0xc01d10f790)
> 	runtime/map.go:853 +0x554 fp=0xc01d10f6a8 sp=0xc01d10f628 pc=0x410db4
> github.com/minio/minio/cmd.(*dataUsageCache).flatten(0xc01d10fe88, 0xa7a5c084cec, 0x0, 0x0, 0x0, 0x0, 0x8b86, 0x9e3, 0x2430, 0x48ce, ...)
> 	github.com/minio/minio/cmd/data-usage-cache.go:342 +0xe9 fp=0xc01d10f870 sp=0xc01d10f6a8 pc=0x16770c9
> github.com/minio/minio/cmd.(*dataUsageCache).dui(0xc01d10fe88, 0x1fd3916, 0x1, 0xc021b98000, 0x2cc0, 0x3733, 0x0, 0x0, 0x0, 0x0, ...)
> 	github.com/minio/minio/cmd/data-usage-cache.go:249 +0xe5 fp=0xc01d10fab8 sp=0xc01d10f870 pc=0x1675f65
> github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3.1()
> 	github.com/minio/minio/cmd/erasure-server-pool.go:443 +0x325 fp=0xc01d10fcf8 sp=0xc01d10fab8 pc=0x18b06a5
> github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage.func3(0xc011b9ee98, 0xc023537e40, 0xc01f548fc0, 0xc021b98000, 0x2cc0, 0x3733, 0xc0017932f0, 0xc0125013e0, 0xc001793310, 0xc001698180)
> 	github.com/minio/minio/cmd/erasure-server-pool.go:462 +0x183 fp=0xc01d10ff90 sp=0xc01d10fcf8 pc=0x18b09c3
> runtime.goexit()
> 	runtime/asm_amd64.s:1374 +0x1 fp=0xc01d10ff98 sp=0xc01d10ff90 pc=0x470701
> created by github.com/minio/minio/cmd.(*erasureServerPools).CrawlAndGetDataUsage
> 	github.com/minio/minio/cmd/erasure-server-pool.go:421 +0x56f
> ```
> 
> It should be fixed in #11620 on `RELEASE.2021-02-24T18-44-45Z` right?

This is fixed @ToLToL

**Comment by ToLToL** [comment]
Ok thanks
